numpyro/distributions/continuous.py

```diff
@@ -1188,6 +1192,141 @@ def infer_shapes(
+    :param ndarray or scipy.sparse.csr_matrix W: symmetric adjacency matrix where 1
+        indicates adjacency between sites and 0 otherwise
+    :param bool is_sparse: whether to use a sparse form of W in calculations (must be True is W is a
+        :class:`scipy.sparse.spmatrix`)
+    """
+
+    arg_constraints = {
+        "loc": constraints.real_vector,
+        "alpha": constraints.open_interval(-1, 1),
+        "tau": constraints.positive,
+        "W": constraints.dependent_property(is_discrete=False, event_dim=2),
+    }
+    support = constraints.real_vector
+    reparametrized_params = [
+        "loc",
+        "alpha",
+        "tau",
+        "W",
+    ]
+
+    def __init__(self, loc, alpha, tau, W, *, is_sparse=False, validate_args=None):
+        if jnp.ndim(loc) == 0:
+            (loc,) = promote_shapes(loc, shape=(1,))
+
+        self.is_sparse = is_sparse
+
+        if self.is_sparse:
+            if not sparse.issparse(W):
+                assert isinstance(W, np.ndarray)
+            self.W = sparse.csr_matrix(W)
+        else:
+            assert not sparse.issparse(W)
+            self.W = W
+
+        batch_shape = lax.broadcast_shapes(
+            jnp.shape(loc)[:-1],
+            jnp.shape(alpha),
+            jnp.shape(tau),
+            jnp.shape(self.W)[:-2],
+        )
+        event_shape = jnp.shape(self.W)[-1:]
+        (self.loc,) = promote_shapes(loc, shape=batch_shape + event_shape)
+        self.alpha, self.tau = promote_shapes(alpha, tau, shape=batch_shape)
+
+        super(CAR, self).__init__(
+            batch_shape=batch_shape,
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+        if self._validate_args:
+            assert (
+                self.W.sum(axis=-1) > 0
+            ), "all sites in adjacency matrix W must have neighbours"
+
+            if sparse.issparse(self.W):
+                assert (self.W != self.W.T).nnz == 0, "W must be symmetric"
+            else:
+                assert np.array_equal(self.W, self.W.T), "W must be symmetric"
+
+    def sample(self, key, sample_shape=()):
+        # TODO: look into a sparse sampling method
+        mvn = MultivariateNormal(self.mean, precision_matrix=self.precision_matrix)
+        return mvn.sample(key, sample_shape=sample_shape)
+
+    @validate_sample
+    def log_prob(self, value):
+        phi = value - self.loc
+        W = self.W
+
+        if self.is_sparse:
+            D = np.asarray(W.sum(axis=-1)).squeeze()
+            D_rsqrt = D ** (-0.5)
+
+            W_scaled = W.multiply(D_rsqrt).multiply(D_rsqrt[:, np.newaxis]).toarray()
+
+            lam = np.linalg.eigvalsh(W_scaled)
+
+            W = BCOO.from_scipy_sparse(W)
+
+        else:
+            D = W.sum(axis=-1)
+            D_rsqrt = D ** (-0.5)
+
+            W_scaled = W * (D_rsqrt * D_rsqrt[:, jnp.newaxis])
+
+            lam = jnp.linalg.eigvalsh(W_scaled)
```

Comment:
Do you want to save the computation as we discussed in several threads? If so, you can do
```
if self.is_sparse:
    ...  # remove lam = ...
else:
    ... # remove lam = ...
if isinstance(W_scaled, np.ndarray):
    lam = np.linalg....
else:
    lam = jnp.linalg...
```

Reply:
I thought we wanted to use numpy for both branches because it will only do the computation once and we only need the calculation done once. Whereas I thought jax.numpy would repeat the calculation at each mcmc step.

So in fact, I should only have `lam = np.linalg.eigvalsh(W_scaled)` once outside of the if-else logic. Do you agree?

Reply:
The tricky thing is if W_scaled is a jax array/tracer, we should use a jax operator for it.

Reply:
The numpy version works on `W_scaled` as jax. The only difference is the output. The jnp version outputs a jax array and the numpy version a numpy array. Equally, the jnp version can take in a numpy array.

If you believe the jax numpy version repeats the calculation at each step, I think we should only use the numpy version. The whole point in the stan article is to calculate the eigenvalues _once beforehand_

Reply:
I see, so np.linalg under jit might not work. Even converting the W_scaled jax array to a numpy array before the operation probably won't work under jit e.g.
```
@jax.jit
def f():
  x = jax.numpy.ones(10)
  x = np.asarray(x)
  return x

f()
```

Is there a workaround to avoid multiple computations of `lam` for the jax case?

Reply:
If we want to avoid computations, we need to treat those values as constant. You just need to provide numpy arrays and use numpy operators on the top of them. The only way we can access values of abstract arrays that I know is to use jax host_callback. But with host_callback, the computation is not cached and more expensive.